### ISSUE TITLE
Geo-location performance 

### ISSUE THREAD

**Issue by andreib1** [arangodb/arangodb#1536]
I have found there are issues with Geo-location when it comes to applying additional filters.  Geo-location does not seem to utilize its index with the rest of the query.

You can see this for yourself if you load test data:

db._useDatabase("_system");
db._createDatabase("WithinTest");
db._useDatabase("WithinTest");
db._create("LatLongs");
for (var i = 0; i < 30000; i++) { db.LatLongs.save({lat: 0, long: 0, number:
i }); }
db.LatLongs.ensureGeoIndex("lat", "long");

And run the following query:

FOR x IN WITHIN(LatLongs, 0, 0, 0)
LIMIT 1
RETURN x._key

On my machine .75 to 1 second to execute for a single result, which implies that it is not streaming results from its index, and that it is enumerating all results before proceeding the query.

**Comment by dylanclewis** [comment]
It seems very strange to me that the limit of 1 has no impact on the performance of the query.

**Comment by dothebart** [comment]
If I get that correctly you put all items to exactly one point (0,0). So you produce an index on lat/long with a [very bad selectivity](http://www.akadia.com/services/ora_index_selectivity.html) . So the index rather harms the performance than improving it and in the end you do an expensive index lookup that results in a full table scan.

**Comment by andreib1** [comment]
A typical use case for us is:
I am standing in a city where many hundreds / thousands of clients are running this software.  
I need to find:  
* Other clients WITHIN a 1 mile radius around me,  
* FILTER by some client attributes,  
* Filtered by a graph lookup,  
* Limit the number to 400,  
* Show me their Client Id  

I acknowledge that the query has bad selectivity.  The problem here though is that the WITHIN query enumerates the result before returning it, and returns the result rather than an index.  This means that the next filter stages have no impact on the amount of documents processed.

**Comment by dothebart** [comment]
Ok, but usually the `WHITHIN()` will not give you 30k documents as in your test, but rather some thousand.
You're right, currently there is no way to combine geo indices with other index operations like filter.
This will change, but not in the near future.

**Comment by andreib1** [comment]
I have a feeling this will be the case with any V8 portions of AQL queries.  I suspect graph queries e.g. GRAPH_NEIGHBORS behave the same, and cannot combine their indexes.  @dothebart Can you confirm? If so this is pretty serious news for anyone who has big data, and the need to combine filtering or limiting with graph/geo.

**Comment by dothebart** [comment]
yes, you only can use joins afterwards to combine the results.

**Comment by jsteemann** [comment]
The performance of the `NEAR()` and `WITHIN()` AQL functions has been improved much in 2.8/devel compared to 2.7, especially for bigger result sets. 
Filters or functions that concern different indexes still are evaluated separately, but there'll be the following improvements (planned) for this:
* 2.8 will provide a native traversal functionality for use within AQL, that will lazily build results and that streams that to further filters
* for a future version > 2.8 we plan to do the same for the geo index functionality. This should also allow building result sets lazily and in a streaming fashion. That should greatly improve the performance for cases such as `WITHIN()/NEAR() LIMIT 1` queries, as currently the functions always build up the entire result. The future geo index API should allow to build only as much results as necessary, so `LIMIT 1` will effectively be applied after the first result and no further documents will need to be retrieved and shipped through the query.

**Comment by dothebart** [comment]
2.8 is available for download now. Get it and enjoy. Closing this issue now, If problems persist, please open a new ticket since this one is way crowded ;-)

**Comment by ra0o0f** [comment]
> for a future version > 2.8 we plan to do the same for the geo index functionality. This should also allow building result sets lazily and in a streaming fashion. That should greatly improve the performance for cases such as WITHIN()/NEAR() LIMIT 1 queries, as currently the functions always build up the entire result. The future geo index API should allow to build only as much results as necessary, so LIMIT 1 will effectively be applied after the first result and no further documents will need to be retrieved and shipped through the query.

@jsteemann is limiting result set with `limit` implemented in geo index functionality?

**Comment by jsteemann** [comment]
@ra0o0f : this has been implemented for 3.2. 

Geo indexes can now be used implicitly when using SORT/FILTER in AQL, without the need to use the special-purpose geo AQL functions `NEAR` or `WITHIN`.

The special purpose `NEAR` AQL function can now be substituted with the following AQL (provided there is a geo index present on the `doc.latitude` and `doc.longitude` attributes):
```
FOR doc in geoSort 
  SORT DISTANCE(doc.latitude, doc.longitude, 0, 0) 
  LIMIT 5 
  RETURN doc
```
`WITHIN` can be substituted with the following AQL:
```
FOR doc in geoFilter 
  FILTER DISTANCE(doc.latitude, doc.longitude, 0, 0) < 2000 
  RETURN doc
```
Compared to using the special purpose AQL functions this approach has the advantage that it is more composable, and will also honor any `LIMIT` values used in the AQL query.

**Comment by ra0o0f** [comment]
@jsteemann Perfect, thanks for the great work
